agent/config/runtime.go

```diff
@@ -1930,3 +1930,16 @@ func isFloat(t reflect.Type) bool { return t.Kind() == reflect.Float32 || t.Kind
 func isComplex(t reflect.Type) bool {
 	return t.Kind() == reflect.Complex64 || t.Kind() == reflect.Complex128
 }
+
+// ApplyDefaultQueryOptions returns a function which will set default values on
+// the options based on the configuration.
+func ApplyDefaultQueryOptions(config *RuntimeConfig) func(options *structs.QueryOptions) {
```

Comment:
Is there a need to keep this a pointer? Otherwise we may need a nil-check on config

Reply:
Good question! I often try not to use pointers and instead use values when possible. I think it would be fine to use a value here, but I opted for a pointer because `RuntimeConfig` is a very large struct, and generally (because it is so large) we have been passing it around by pointer. I think we expect to only call this method on startup, so using a non-pointer might also be ok.

Regarding the nil-check, if this value were to come from a user then I think that would be very important. But the user can't directly set `RuntimeConfig`. It's expected the agent startup will always create one of these, and there should be no way that it could be nil.

So the only way for `RuntimeConfig` to be `nil` would be a "programming error". A panic is an appropriate way of handling programming errors because we expect our test suite to catch them. If we were to check for the nil, we'd either have to change the signature and return an error from this function, or we'd create a custom panic with more detail.

I suspect a detailed panic wouldn't really help here, since a nil pointer should hopefully be easy to debug. Changing the function signature to return an error would imply a runtime error, but there is no possible runtime error here (only programming errors). I generally try to avoid adding error returns unless there is a possible runtime error because it can force all the callers to also handle the error (which can be a distraction if the only cause is a possible programming error).

I think passing a non-pointer value could in some cases be worse. If we were to pass a zero value for `RuntimeConfig` it would force `MaxQueryTime` to be 0 always. Hopefully that kind of error would be noticed by tests, but a panic should be a lot more obvious.

In short, I think that's an excellent call out (checking for nil), but in this case I'm tempted to say the current implementation is reasonable given that we always expect a non-nil `RuntimeConfig`.

That said, it would be good to document these assumptions in the function docstring, so I will add a line about "RuntimeConfig must not be nil".

Reply:
Thanks for the detailed answer! I definitely got more than I asked for :D 
Your point about programming error vs runtime is really interesting - I've never thought of it that way and have seen my fair share of "this change touches a billion files because something down the stack now returns an error"